### ISSUE TITLE
Document best practices for collaboration

### ISSUE THREAD

**Issue by omaciel** [PulpQE/pulp-smash#131]
Let's create a wiki/document describing 'the rules of engagement', best practices for collaboration with this project. What do we expect from pull requests, issues, etc created by those who want to collaborate with the project, as well as what they should expect from us.

**Comment by asmacdo** [comment]
Can this live in the official docs? If not, then lets remove what is there and put a link. It would be most helpful if all the info needed is in only one place.

**Comment by Ichimonji10** [comment]
I like the idea of putting this in the docs. How about Pulp Smash → About → [Contributing](http://pulp-smash.readthedocs.org/en/latest/about.html#contributing)?

Here's a brief proposal. We can continue submitting pull requests as we currently do, but the standard for pull requests would be as so:

* A pull request should pass all automated linters. Travis already checks this automatically, but this can also be done locally with the various make targets.
* The tests added by a pull request should pass. The various functions provided by [`pulp_smash.selectors`](http://pulp-smash.readthedocs.org/en/latest/api/pulp_smash.selectors.html) are of special use in this regard.
* A pull request should target a single issue. For example, if a given pull request both adds tests and fixes a bug in a utility function, those are two separate concerns, and separate pull requests can be submitted for each. As a side effect, this means that each pull request should contain one commit.

If a pull request satisfies those requirements, and if there are no other outstanding issues, [1] then the reviewer can add a second commit addressing whatever concerns or nitpicks they have, rebase the pair of commits on master, verify the sanity of the result (Does the code still satisfy the linters? Do the affected tests still pass? Do the pair of commits target a single issue?), and merge.

* Commit authorship is preserved. Both of the individuals involved in adding a feature will have their work listed.
* Commit integrity is preserved. Commits can reflect the actual output of individuals, instead of being tweaked invisibly.
* The "linters must pass" requirement is currently automated, and the "tests must pass" requirement can be automated. That's two of three.

There is one downside to this approach that sticks out to me, and it's that the commit history will be less straightforward. This makes using `revert`, `bisect`, `log`, and all manner of other tools harder to do. However, two commits are still pretty reasonable. (We can try experimenting with `git merge --no-ff` when adding submitter/reviewer pairs of commits. That might make the pairs of commits more clear, but the full implications of that approach aren't clear to me yet.)

[1] What qualifies as an "outstanding issue?" Examples might include:

* A commit adds a 10MB binary blob to the repository without exceedingly good reason. Remember that git keeps all data forever, whether or not it is used.
* A commit adds a test that makes dozens of concurrent requests to a publicly available service such as docker hub.

This "outstanding issue" clause is intentionally vague. There's just now way that all issues can be foreseen, and leaving subjective wiggle room seems wise.

**Comment by mhrivnak** [comment]
LGTM. For the section on outstanding issues, you might add some of these thoughts:

- missing unit tests
- not adhering to generally accepted best practices that are difficult to catch with a linter, such as DRY
- anything that would normally be required by a pull request against pulp itself

You may or may not want to add those specific items, but the point is to show that you can request reasonable changes in cases that don't involve such obviously bad ideas as the ones you already cited.

**Comment by omaciel** [comment]
So the tl;dr is:

* A contributor sends a Pull Request addressing a unique, single task.
  * a task could be one single automated test, or many
  * a task could be a fix to an existing piece of the framework, required to support future automated tests
  * therefore, we would like to keep one pull request per unique task
* If the pull request passes Travis (aka lints):
  * If there are only 'cosmetic' nitpicks related to line alignments, etc:
    * The contributor is free to move on
    * Nitpicks will be handled by the reviewer and a second commit will be added to the pull request
  * Else:
    * The reviewer will provide valuable feedback related to the actual test or code submitted that should be addressed, and the cycle repeats
* Else:
  * The contributor needs to address the issues raised by Travis and rebase/resubmit

**Comment by Ichimonji10** [comment]
> For the section on outstanding issues, you might add some of these thoughts:
>
> - missing unit tests

Ahh. So if a pull request states "test foo" but fails to test foo, then I can kick it back to the submitter. That makes sense.

> - not adhering to generally accepted best practices that are difficult to catch with a linter, such as DRY

Hehe. I think that my ideas on crafting reusable code and so on may differ from what the contributor's thoughts to a frustrating extent. It might be best to say "your pull request is more likely to be reviewed and merged in a prompt manner if it exceeds the three requirements listed above. For example, consider ensuring that your code is DRY, matches existing formatting conventions, is organized into easy-to-read blocks, has isolated assertions, is thoughtfully documented and has a [great](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) commit [message](http://chris.beams.io/posts/git-commit/). The sky is the limit, and effort is appreciated." Basically - let's encourage people to reach high, without stepping back into our current situation of frustratingly hard reviews.

**Comment by Ichimonji10** [comment]
@omaciel I'm thinking of having a lower bar than that for getting commits merged, and increasing the scope of what a reviewer might do in their follow-up commit. I really do want to make sure that the requirements for having a pull request considered and eventually merged are clearly laid out and reasonably low, so as to avoid stepping back into our current situation of having frustratingly hard code reviews.

This can lead to a perverse incentive where a submitter meets only the minimum requirements for a pull request and puts a large burden on the pulp smash gatekeepers. Stating "you pull request is more likely to be reviewed merged in a prompt manner if ..." puts that incentive back in place. It gently encourages submitters to reach for the sky by making clear that reviewers have limited time and resources.

**Comment by asmacdo** [comment]
I worry about the model where contributors meet a set of minimum standards and the reviewer is responsible to clean it up for higher standards. At the moment, you are the primary reviewer and commiter, but if you are out or busy and pulp-smash tests need to be merged, I can't justify spending my time rewriting other contributors' commit messages or rewriting code for personal style preferences.

I think that leaving nitpicks to the reviewer is fine, but I want to make sure that those nitpicks are up to the reviewer, not a second much more stringent set of rules.

**Comment by rbarlow** [comment]
I don't think that one commit per pull request should be a hard rule. We do usually do that in Pulp and it is most often the best course, but occasionally it does make sense to have two atomic commits that need to go in together. I would reword that bit to just say that commits should be atomic and leave pull requests out of it.

**Comment by Ichimonji10** [comment]
> I worry about the model where contributors meet a set of minimum standards and the reviewer is responsible to clean it up for higher standards.

Under the current model, I ask contributors to meet high standards, and it's failing spectacularly. This proposal changes things so that contributors are asked to meet lower standards, and are merely incentivized to reach higher. A third option is to just... give up.

> I can't justify spending my time rewriting other contributors' commit messages

That's not part of the proposal.

> or rewriting code for personal style preferences.

Code is read far more often than it is written. I don't think it's a waste of time to ensure that code is legible.

Also, there's much more to consider than style nitpicks. I also deal with cases where a pull request introduces unnecessary or worrisome levels of interface complexity; where inheritance and code re-use is badly implemented; and so on.

**Comment by mhrivnak** [comment]
Let's be careful to keep this discussion in perspective. We want the same things. Everyone involved wants code to be legible. Nobody is arguing in favor of complexity or bad implementations.

But these are subjective measurements whose long-term value is difficult to quantify, and that is the reason for the proposed arrangement.

As you invest more time and energy in moving those measurements in a positive direction, the incremental return on your investment diminishes. Different teams and individuals have to make their own conclusions about how far to go, and at what point the value obtained no longer justifies the expense. The differing needs and circumstances of each project will naturally lead to different conclusions.

The smash project wants to invest more heavily than the pulp project does. In my experience, both projects have a desired level of investment that is higher than what we find in the general python OSS community. That's fine and admirable, and everyone can support smash's desired standards. The proposed agreement allows external contributors to submit PRs that meet a reasonable and generally agreeable set of standards that are likely to be compatible with most contributors' expectations, while allowing smash maintainers to provide extra lift at the end to bring PRs up to smash's higher-than-average standards. The pulp team has also done this with many community contributions, and we've found it to be a successful arrangement.

The concern from @asmacdo is that as an external contributor who also has been asked to help with PR review, he does not want to be responsible for providing that tail-end lift. I think that's very reasonable, and is likely agreeable by all. External reviewers should focus on areas where the author will be asked to make changes.

**Comment by Ichimonji10** [comment]
> The concern from @asmacdo is that as an external contributor who also has been asked to help with PR review, he does not want to be responsible for providing that tail-end lift. I think that's very reasonable, and is likely agreeable by all. External reviewers should focus on areas where the author will be asked to make changes.

That makes sense to me. :+1:

**Comment by Ichimonji10** [comment]
[Issue and Pull Request Templates](https://github.com/blog/2111-issue-and-pull-request-templates)
